repair: harden effective replication map

bhalevy opened the pull request:
As described in #11993 per-shard repair_info instances get the effective_replication_map on their own with no centralized synchronization.

This series ensures that the effective replication maps used by repair (and other associated structures like the token metadata and topology) are all in sync with the one used to initiate the repair operation.

While at at, the series includes other cleanups in this area in repair and view that are not fixes as the calls happen in synchronous functions that do not yield.

Fixes #11993

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3012/

avikivity commented on db/view/view.cc:
```diff
@@ -1391,9 +1391,9 @@ static std::optional<gms::inet_address>
-    auto& topology = service::get_local_storage_proxy().get_token_metadata_ptr()->get_topology();
     auto& ks = db.find_keyspace(keyspace_name);
     auto erm = ks.get_effective_replication_map();
+    auto& topology = erm->get_token_metadata_ptr()->get_topology();
```
The patch is okay, but since we drop erm after returning the result, it could be out of date when used. Probably it will be fixed later on in the series.

bhalevy replied:
not in this series, that's focused on repair.
this is a localized cleanup to just get the topology from the erm that's is already retrieved here.

avikivity replied:
I think we need to decouple mv updates from storage_proxy, or we'll get into a difficult circular dependency problem. That is, the mv update is a new sub-transaction with its own erm.

avikivity commented on repair/repair.cc:
```diff
@@ -562,6 +562,7 @@ repair_info::repair_info(repair_service& repair,
     , gossiper(repair.get_gossiper())
     , sharder(get_sharder_for_tables(db, keyspace_, table_ids_))
     , keyspace(keyspace_)
+    , erm(db.local().find_keyspace(keyspace).get_effective_replication_map())
```
This is good, but forces us to consider several paths:

1. we keep erm and block topology changes until all its users are gone - so repair blocks topology changes
2. we add erm->transitioning_away_from() and sample it regularly during repair, failing or restarting repair if it happens. This allows topology changes to preempt repair
3. some mix where we complete a partial repair and then restart it using the new topology

Anyway, this is okay for now, but we rely on the operator not to change topology while repair is running.

bhalevy replied:
Cc @asias @tgrabiec

avikivity commented on repair/repair.cc:
```diff
@@ -1191,12 +1194,18 @@ int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring
+                    data_centers = options.data_centers, hosts = options.hosts, ignore_nodes, ring_version] (repair_service& local_repair) mutable {
+                auto erm = local_repair.get_db().local().find_keyspace(keyspace).get_effective_replication_map();
+                auto local_ring_version = erm->get_token_metadata().get_ring_version();
+                if (local_ring_version != ring_version) {
```
This is good for now (and an improvement) but probably insufficient. For example, we can get a different erm if we have replication strategy parameter changes.

bhalevy replied:
It's a bit subtle but the schema change is processed by storage_service::on_update_keyspace that calculates a new e_r_m and when applying it we recalculate the token_metadata, although it's didn't explicitly change, and give it a new version, so nowadays the token_metadata::ring_version covers both the token_metadata and the effective_replication_map.  Topology changes should also cause recalculation of this bundle.
Maybe we need to define a `global_effective_replication_map` that will get the e_r_m on all shards (with grabbing the shared_token_metadata lock so it won't change) and then we can pass it around across shards.

avikivity commented on repair/repair.cc:
```diff
@@ -1285,15 +1296,21 @@ future<> repair_service::sync_data_using_repair(
+                auto local_ring_version = erm->get_token_metadata().get_ring_version();
+                if (local_ring_version != ring_version) {
+                    return make_exception_future<>(std::runtime_error("Ring version changed"));
+                }
```
Perhaps we'll want to make a sharded<erm>.

bhalevy replied:
~= see above.

avikivity commented:
@asias review please

bhalevy commented on repair/repair.cc:
```diff
@@ -1034,7 +1034,8 @@ int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring
     } else {
-        ranges = db.get_keyspace_local_ranges(keyspace);
+        // get keyspace local ranges
+        ranges = erm->get_ranges(utils::fb_utilities::get_broadcast_address());
```
very much so

bhalevy commented:
In v2 ([6012230](https://github.com/scylladb/scylladb/commit/6012230eb4bd2f5734dd360a8076e7705f2bc667)):
- shared_token_metadata: get_lock is const
- effective_replication_map: add global_effective_replication_map
- repair: futurize do_repair_start
- repair: coroutinize sync_data_using_repair
- repair: pass effective_replication_map down to repair_info
  - use global_effective_replication_map
- repair: pass erm down to get_hosts_participating_in_repair and get_neighbors
  - pass const locator::effective_replica_map& to sync functions rather than *_ptr

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3029/

asias commented on locator/abstract_replication_strategy.cc:
```diff
@@ -468,6 +470,32 @@ void effective_replication_map_factory::submit_background_work(future<> fut) {
+            _erms[shard] = co_await sharded_db.invoke_on(shard, [keyspace_name, ring_version] (const replica::database& db) {
+                const auto& ks = db.find_keyspace(keyspace_name);
+                auto erm = ks.get_effective_replication_map();
+                auto local_ring_version = erm->get_token_metadata().get_ring_version();
```
Can we 100% sure that all the share will get the same ring version at any point in time even during the erm is replicated to all shards?

bhalevy replied:
Yes.  This is the way the token_metadata + replication maps are updated on all shards.
We first calculate and update the copy on shard 0 and then clone them onto all shards, including their ring version, under the shared_token_metadata lock.  Hence the on_internal_error here.

bhalevy replied:
I added a comment in v3 describing the mechanism.

asias commented on repair/repair.cc:
```diff
@@ -1042,14 +1025,19 @@ int repair_service::do_repair_start(sstring keyspace, std::unordered_map<sstring
+            // get_primary_ranges_within_dc() is similar to get_primary_ranges(),
+            // but instead of each range being assigned just one primary owner
+            // across the entire cluster, here each range is assigned a primary
+            // owner in each of the clusters.
```
typo: in each of the DCs.

bhalevy replied:
yeah, I see it was preexisting and I merely moved the comment here, but I'll fix it while at it.

bhalevy replied:
fixed in v3

asias commented:
Looks good to me.

bhalevy commented:
In v3 [53fdf75](https://github.com/scylladb/scylladb/commit/53fdf75cf9b404159d230a1b1a81bed3b221f289)):
- rebased
- fixed mistake in comment regarding get_primary_ranges_within_dc
- added comment documenting the algorithm and invariant for global_effective_replication_map.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3034/
